### PR TITLE
Typehinted Route dir

### PR DESCRIPTION
**Pull request by will2877**
## Changelog
Added type hints to all files in the Route directory

### CODE REVIEW

File: src/Route/AdminPoolLoader.php
Comment (VincentLanglet) [inline src/Route/AdminPoolLoader.php:67]: @sonata-project/contributors Should decide a CS rule about this ?

Comment (will2877) [reply src/Route/AdminPoolLoader.php:67]: It was suggested by PHPStorm, that is why I adjusted it.
Should I change it back?

File: src/Route/RouteCollection.php
Comment (VincentLanglet) [inline src/Route/RouteCollection.php:178]:
```suggestion
```
This can be omitted since we never use `@return $this`.

File: src/Route/RouteGeneratorInterface.php
Comment (VincentLanglet) [inline src/Route/RouteGeneratorInterface.php:29]:
```suggestion
    public function generateUrl(AdminInterface $admin, string $name, array $parameters = [], int $referenceType = UrlGeneratorInterface::ABSOLUTE_PATH): string;
```

File: src/Route/RouteGeneratorInterface.php
Comment (VincentLanglet) [inline src/Route/RouteGeneratorInterface.php:36]:
```suggestion
    public function generateMenuUrl(AdminInterface $admin, string $name, array $parameters = [], int 
$referenceType = UrlGeneratorInterface::ABSOLUTE_PATH): array;
```

File: src/Route/RouteGeneratorInterface.php
Comment (VincentLanglet) [inline src/Route/RouteGeneratorInterface.php:43]:
```suggestion
    public function generate(string $name, array $parameters = [], int $referenceType = UrlGeneratorInterface::ABSOLUTE_PATH): string;
```

File: src/Route/RouteGeneratorInterface.php
Comment (VincentLanglet) [inline src/Route/RouteGeneratorInterface.php:48]:
```suggestion
    public function hasAdminRoute(AdminInterface $admin, string $name): bool;
```

File: src/Route/RoutesCache.php
Comment (VincentLanglet) [inline src/Route/RoutesCache.php:45]: You can add typehint to the construct

File: src/Route/AdminPoolLoader.php
Comment (VincentLanglet) [reply src/Route/AdminPoolLoader.php:67]: I have no opinion on this.

But I would recommend to revert it indeed
- to avoid potential conflict
- avoid delaying this PR because of a discussion about this change
- it's better to have this change made automatically

I looked if Phpcs https://github.com/FriendsOfPHP/PHP-CS-Fixer provides a rule about this change. And yes:
Seems like it's the `global_namespace_import` rule.

If we update our config (https://github.com/sonata-project/dev-kit/blob/master/templates/project/.php_cs.dist) we could then have this change made automatically on every repository and it would give us a real consistency.

Do you want to try a PR on dev-kit ? ;)

File: src/Route/AdminPoolLoader.php
Comment (will2877) [reply src/Route/AdminPoolLoader.php]: This is incompatible with the Interface and causing the tests to fail.

File: src/Route/AdminPoolLoader.php
Comment (VincentLanglet) [inline src/Route/AdminPoolLoader.php:51]: You'll get a phpcs issue here (and same for other changes).

Even if `?` is not required, a phpcs rule is requiring it when `null` is used as a default value.

File: src/Route/AdminPoolLoader.php
Comment (VincentLanglet) [reply src/Route/AdminPoolLoader.php]: Nevermind I opened the PR https://github.com/sonata-project/dev-kit/pull/1170 @will2877

File: src/Route/AdminPoolLoader.php
Comment (will2877) [reply src/Route/AdminPoolLoader.php:51]: I saw this, what should I do?
If I add the `?` the Interface is incompatible, else the phpcs fails.

Comment (VincentLanglet) [reply src/Route/AdminPoolLoader.php:51]: I would be really surprised if adding the `?` is incompatible with the Interface.

`?string $type = null` and `string $type = null` is the same thing.

File: src/Route/AdminPoolLoader.php
Comment (VincentLanglet) [reply src/Route/AdminPoolLoader.php]: Return type can always be added.
But indeed, the interface doesn't have a phpdoc about the return type, which means, `void` seems expected.

Maybe the return type should be `void` and we should change something in the implementation.
CC @sonata-project/contributors @jordisala1991

**Review by VincentLanglet — COMMENTED** [review]
And there is still my comment on RouteCollection.

Otherwise it seems perfect.

File: src/Route/AdminPoolLoader.php
Comment (VincentLanglet) [inline src/Route/AdminPoolLoader.php:55]:
```suggestion
     * @param string|null $type
```

File: src/Route/AdminPoolLoader.php
Comment (VincentLanglet) [inline src/Route/AdminPoolLoader.php:68]:
```suggestion
     * @param string|null $type
```

**Comment by VincentLanglet** [comment]
@will2877 You added phpdoc for the return type of  `supports` and `load`, but you can directly write
```
supports($resource, $type = null): bool
load($resource, $type = null): SymfonyRouteCollection
```

It's still compatible with the interface (you can't add param type hint, but you can add return typehint).

But @sonata-project/contributors, I have an issue with the `AdminLoader::load` method, in Symfony it seems like it returns nothing (https://github.com/symfony/config/blob/5.x/Loader/LoaderInterface.php) ; so I wonder if there is not something to change to respect the signature
```
load($resource, $type = null): void
```
instead

**Comment by VincentLanglet** [comment]
@OskarStark @jordisala1991 Any idea why `flintci/pr` is failing ?
Shouldn't this check be removed since we have `Lint / PHP-CS-Fixer (pull_request)` ?

File: src/Route/AdminPoolLoader.php
Comment (will2877) [reply src/Route/AdminPoolLoader.php]: This is a CS Fixer adjustment.

File: src/Route/AdminPoolLoader.php
Comment (will2877) [reply src/Route/AdminPoolLoader.php]: This is a CS Fixer adjustment.

**Comment by VincentLanglet** [comment]
@will2877 There is something wrong with your param typehint
```
* @param string $type     | null
```

I expect
```
* @param string|null $type
```

You're currently always adding the `| null` after `$type`, so it's considered as a comment.

File: src/Route/AdminPoolLoader.php
Comment (core23) [inline src/Route/AdminPoolLoader.php:55]:
```suggestion
     * @param string|null $type
```

File: src/Route/AdminPoolLoader.php
Comment (core23) [inline src/Route/AdminPoolLoader.php:66]:
```suggestion
     * @param string|null $type
```

**Review by core23 — CHANGES_REQUESTED** [review]

File: src/Route/AdminPoolLoader.php
Comment (VincentLanglet) [inline src/Route/AdminPoolLoader.php:54]: Are you sure it's a string ?

In the Symfony interface I see
```
@param mixed       $resource
```

File: src/Route/AdminPoolLoader.php
Comment (VincentLanglet) [inline src/Route/AdminPoolLoader.php:65]: Same

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

**Review by VincentLanglet — APPROVED** [review]

File: src/Route/AdminPoolLoader.php
Comment (franmomu) [inline src/Route/AdminPoolLoader.php:52]: IMO we should find another comment instead of `NEXT_MAJOR` when it's about dropping support of packages because it could be confusing, moreover, adding them in `master`, related: https://github.com/sonata-project/SonataMediaBundle/pull/1814#discussion_r492047513.

It would be nice to include the standard name of the specific package in the message, like in this case `symfony/config`.

Comment (will2877) [reply src/Route/AdminPoolLoader.php:52]: @franmomu 
What do you suggest to put there?
a todo?

Comment (VincentLanglet) [reply src/Route/AdminPoolLoader.php:52]: @franmomu We're currently using this comment elsewhere.

Moreover this change will need a next major since its BC break.

We shouldnt block this PR for this, and we could discuss about this on devkit, wdyt ?

Comment (franmomu) [reply src/Route/AdminPoolLoader.php:52]: I'm not blocking this PR (not requesting changes), just a comment because I feel we will have to come back to change these comments.

This won't be BC break since the class is final.

**Comment by VincentLanglet** [comment]
@core23 You had requested changes. Please review again :)

**Comment by VincentLanglet** [comment]
> @core23 You had requested changes. Please review again :)

Or if you can take a look @greg0ire / @wbloszyk

**Review by core23 — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Thanks @will2877

**Pull request merged**
